File: test-suite-thymeleaf-fieldset/src/test/java/io/micronaut/views/fields/thymeleaf/BookControllerTest.java
package io.micronaut.views.fields.thymeleaf;

import io.micronaut.context.annotation.Property;
import io.micronaut.core.type.Argument;
import io.micronaut.http.HttpRequest;
import io.micronaut.http.HttpStatus;
import io.micronaut.http.client.BlockingHttpClient;
import io.micronaut.http.client.HttpClient;
import io.micronaut.http.client.annotation.Client;
import io.micronaut.http.client.exceptions.HttpClientResponseException;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import org.junit.jupiter.api.Test;

import java.util.Optional;

import static io.micronaut.views.fields.thymeleaf.TestUtils.formPost;
import static io.micronaut.views.fields.thymeleaf.TestUtils.htmlGet;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

@Property(name = "datasources.default.password", value = "")
@Property(name = "datasources.default.dialect", value = "H2")
@Property(name = "datasources.default.schema-generate", value = "CREATE_DROP")
@Property(name = "datasources.default.url", value = "jdbc:h2:mem:devDb;LOCK_TIMEOUT=10000;DB_CLOSE_ON_EXIT=FALSE")
@Property(name = "datasources.default.username", value = "sa")
@Property(name = "datasources.default.driver-class-name=org.h2.Driver")
@Property(name = "spec.name", value = "BookControllerTest")
@MicronautTest
class BookControllerTest {

    @Test
    void itIsPossibleToSave(@Client("/") HttpClient httpClient) {
        BlockingHttpClient client = httpClient.toBlocking();
        String html = assertDoesNotThrow(() -> client.retrieve(htmlGet("/books/list")));
        assertTrue(html.contains("<!DOCTYPE html>"));
        assertFalse(html.contains("<li>"));

        html = assertDoesNotThrow(() -> client.retrieve(htmlGet("/books/create")));
        assertTrue(html.contains("""
                <form action="/books/save" method="post">\
                <div class="mb-3">\
                <label for="title" class="form-label">Title</label>\
                <input type="text" name="title" value="" id="title" minlength="2" maxlength="255" class="form-control" required="required"/>\
                </div>\
                <div class="mb-3">\
                <label for="pages" class="form-label">Pages</label>\
                <input type="number" name="pages" value="" id="pages" min="1" max="21450" class="form-control" required="required"/>\
                </div>\
                <input type="submit" value="Submit" class="btn btn-primary"/>\
                </form>"""));
        assertFalse(html.contains("<li>"));

        html = assertDoesNotThrow(() -> client.retrieve(formPost("/books/save", "title=Building Microservices&pages=120")));
        assertTrue(html.contains("<!DOCTYPE html>"));
        assertTrue(html.contains("<li><span>Building Microservices</span><span>120</span></li>"));
    }

    @Test
    void formIsValidated(@Client("/") HttpClient httpClient) {
        BlockingHttpClient client = httpClient.toBlocking();
        Argument<String> okResponse = Argument.of(String.class);
        Argument<String> failureResponse = Argument.of(String.class);
        HttpRequest<?> request = formPost("/books/save", "title=Building Microservices&pages=0");

        HttpClientResponseException e = assertThrows(HttpClientResponseException.class, () -> client.retrieve(request, okResponse, failureResponse));
        assertEquals(HttpStatus.UNPROCESSABLE_ENTITY, e.getStatus());
        Optional<String> htmlOptional = e.getResponse().getBody(String.class);
        assertTrue(htmlOptional.isPresent());
        String html = htmlOptional.get();
        assertTrue(html.contains("""
                <form action="/books/save" method="post">\
                <div class="mb-3">\
                <label for="title" class="form-label">Title</label>\
                <input type="text" name="title" value="Building Microservices" id="title" minlength="2" maxlength="255" class="form-control" required="required"/>\
                </div>\
                <div class="mb-3">\
                <label for="pages" class="form-label">Pages</label><input type="number" name="pages" value="0" id="pages" min="1" max="21450" class="form-control is-invalid" aria-describedby="pagesValidationServerFeedback" required="required"/>\
                <div id="pagesValidationServerFeedback" class="invalid-feedback">must be greater than or equal to 1</div>\
                </div>\
                <input type="submit" value="Submit" class="btn btn-primary"/>\
                </form>"""));

    }
}

File: test-suite-python/src/test/python/micronaut/views/docs/fieldset/BookControllerTest.py
from typing import Annotated

from jakarta.inject import Inject
from java.lang import String
from micronaut.context.annotation import Property
from micronaut.http import HttpRequest, HttpStatus, MediaType
from micronaut.http.client import HttpClient
from micronaut.http.client.annotation import Client
from micronaut.http.client.exceptions import HttpClientResponseException
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test


@Property(name="micronaut.views.soy.enabled", value="false")
@Property(name="micronaut.security.enabled", value="false")
@Property(name="spec.name", value="BookControllerTest")
@MicronautTest
class BookControllerTest:
    httpClient: Annotated[HttpClient, Inject, Client("/")]

    @Test
    def test_the_form_is_generated_validated_and_saved(self):
        client = self.httpClient.toBlocking()
        html = client.retrieve(self.html_get("/books/create"))
        assert ('<form action="/books/save" method="post">'
                '<div class="mb-3">'
                '<label for="title" class="form-label">Title</label>'
                '<input type="text" name="title" value="" id="title" minlength="2" maxlength="255" class="form-control" required="required"/>'
                '</div>'
                '<div class="mb-3">'
                '<label for="pages" class="form-label">Pages</label>'
                '<input type="number" name="pages" value="" id="pages" min="1" max="21450" class="form-control" required="required"/>'
                '</div>'
                '<input type="submit" value="Submit" class="btn btn-primary"/>'
                '</form>') in html

        try:
            client.retrieve(self.form_post("/books/save", "title=Building Microservices&pages=0"))
            assert False, "HttpClientResponseException expected"
        except HttpClientResponseException as e:
            assert e.getStatus() == HttpStatus.UNPROCESSABLE_ENTITY
            error_html = e.getResponse().getBody(String).orElseThrow()
            assert '<input type="number" name="pages" value="0" id="pages" min="1" max="21450" class="form-control is-invalid" aria-describedby="pagesValidationServerFeedback" required="required"/>' in error_html
            assert '<div id="pagesValidationServerFeedback" class="invalid-feedback">must be greater than or equal to 1</div>' in error_html

        html = client.retrieve(self.html_get("/books/list"))
        assert "<li>" not in html

        html = client.retrieve(self.form_post("/books/save", "title=Building Microservices&pages=120"))
        assert "<li><span>Building Microservices</span><span>120</span></li>" in html

    def html_get(self, path: str):
        return HttpRequest.GET(path).accept(MediaType.TEXT_HTML)

    def form_post(self, path: str, body: str):
        return HttpRequest.POST(path, body).contentType(MediaType.APPLICATION_FORM_URLENCODED).accept(MediaType.TEXT_HTML)
